src/library/searchquery.cpp

```diff
@@ -471,6 +471,217 @@ double DurationFilterNode::parse(const QString& arg, bool* ok) {
+                // decimals in tracks table / BPM widgets)
+                // TODO What about -bpm: ? ExactNot
+                m_matchMode = MatchMode::Exact;
+            } else {
+                m_matchMode = MatchMode::Operator;
+            }
+        }
+    } else {
+        if (fuzzy) {
+            // fuzzy can't be combined with range
+            return;
+        }
+        // Test if this is a valid range query
+        QStringList rangeArgs = argument.split("-");
+        if (rangeArgs.length() == 2) {
+            bool lowOk = false;
+            bool highOk = false;
+            double rangeLower = rangeArgs[0].toDouble(&lowOk);
+            double rangeUpper = rangeArgs[1].toDouble(&highOk);
+
+            if (lowOk && highOk && rangeLower <= rangeUpper) {
+                // Assign values now to avoid moving bounds out of scope
+                m_matchMode = MatchMode::Range;
+                m_rangeLower = rangeLower;
+                m_rangeUpper = rangeUpper;
+                return;
+            }
+        }
+        // m_matchMode is already Invalid.
+        return;
+    }
+
+    // Build/prepare match functions
+    switch (m_matchMode) {
+    case MatchMode::Exact: {
+        std::tie(m_rangeLower, m_rangeUpper) = rangeFromTrailingDecimal(bpm);
+        break;
+    }
+    case MatchMode::Fuzzy: {
+        m_rangeLower = floor((1 - s_relativeRange) * bpm);
+        m_rangeUpper = ceil((1 + s_relativeRange) * bpm);
+        break;
+    }
+    case MatchMode::HalveDouble: {
+        std::tie(m_rangeLower, m_rangeUpper) = rangeFromTrailingDecimal(bpm);
+        m_bpmHalfLower = floor(bpm / 2);
+        m_bpmHalfUpper = ceil(bpm / 2);
+        m_bpmDoubleLower = floor(bpm * 2);
+        m_bpmDoubleUpper = ceil(bpm * 2);
+        break;
+    }
+    case MatchMode::Operator: {
+        m_bpm = bpm;
+        break;
+    }
+    default:
+        return;
+    }
+}
+
+bool BpmFilterNode::match(const TrackPointer& pTrack) const {
+    double value = pTrack->getBpm();
+
+    switch (m_matchMode) {
+    case MatchMode::Null: {
+        return value == 0.0;
+    }
+    case MatchMode::Exact:
+    case MatchMode::Fuzzy:
+    case MatchMode::Range: {
+        return value >= m_rangeLower && value <= m_rangeUpper;
+    }
+    case MatchMode::HalveDouble: {
+        return (value >= m_rangeLower && value <= m_rangeUpper) ||
+                (value >= m_bpmHalfLower && value <= m_bpmHalfUpper) ||
+                (value >= m_bpmDoubleLower && value <= m_bpmDoubleUpper);
+    }
+    case MatchMode::Operator: {
+        return (m_operator == "=" && value == m_bpm) ||
+                (m_operator == "<" && value < m_bpm) ||
+                (m_operator == ">" && value > m_bpm) ||
+                (m_operator == "<=" && value <= m_bpm) ||
+                (m_operator == ">=" && value >= m_bpm);
+    }
+    default: // e.g. MatchMode::Invalid
+        // Show no results to indicate the query is invalid.
```

Comment:
this can be a debug assert -- in a release build we can just silently fail, but during development we'd want to know it was happening

Reply:
MatchMode::Invalid is not for programmming errors but for malformed **typed** queries (user error), for example
fuzzy range ~bpm:123-125
fuzzy operator ~bpm:<=95